jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBufferPool.java

```diff
@@ -0,0 +1,154 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+import java.util.function.Consumer;
+
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Pool;
+import org.eclipse.jetty.util.annotation.ManagedAttribute;
+import org.eclipse.jetty.util.annotation.ManagedObject;
+
+@ManagedObject
+public class RetainableByteBufferPool implements MemoryPool<RetainableByteBuffer>
+{
+    private final AtomicReferenceArray<Pool<RetainableByteBuffer>> _direct;
+    private final AtomicReferenceArray<Pool<RetainableByteBuffer>> _indirect;
+    private final int _factor;
+    private final int _maxBucketSize;
+    private final int _minCapacity;
+
+    public RetainableByteBufferPool()
+    {
+        this(0, 1024, 65536, Integer.MAX_VALUE);
+    }
+
+    public RetainableByteBufferPool(int minCapacity, int factor, int maxCapacity, int maxBucketSize)
+    {
+        _factor = factor <= 0 ? 1024 : factor;
+        _maxBucketSize = maxBucketSize;
+        if (minCapacity <= 0)
+            minCapacity = 0;
+        _minCapacity = minCapacity;
+        if (maxCapacity <= 0)
+            maxCapacity = 64 * 1024;
+        if ((maxCapacity % _factor) != 0 || _factor >= maxCapacity)
+            throw new IllegalArgumentException("The capacity factor must be a divisor of maxCapacity");
+
+        int length = maxCapacity / _factor;
+
+        _direct = new AtomicReferenceArray<>(length);
+        _indirect = new AtomicReferenceArray<>(length);
```

Comment:
What is the saving making these arrays atomic?  Would it not be better to allocate all at once?      If the thinking is that most buckets will not be used, so we will not create most of those sub arrays, then we have to be careful we don't expose ourselves to DOS attacks where an attacker somehow forces all the buckets to be allocated. If we can't guarantee that we can prevent all the buckets being allocated then we should allocate them all up front so we know we have space.

Perhaps we need a better than a linear slice to allocate buckets.  If we use 8k and 16k buckets, will we really ever have need of 9-15k buckets?   What about a simple log bucket calculation so buckets go 1K, 2K, 4K, 8K, 16K, 32K & 64K - so 7 buckets for a fully populated pool. 

Hmm an empty bucket is just an empty COW array list right??? So not expensive to fully allocate all buckets even if linear... but I still think there is a DOS  issue with linear allocated buckets, so log buckets are better.

Reply:
This mechanism is almost identical to the `ArrayByteBuffer` one. The only difference is that I used atomic ref arrays instead of plain arrays to avoid creating non-empty buckets that could be dropped as garbage. 

The current algorithm served us well until now, so if we want to modify it, I think that should be done as a separate PR. For instance, using log bucket sizes already opens questions about the SSL case which uses 17K buffers. And I do not think we're open for DoS attacks as AFAIK the clients have no control over the bucket sizes.

Though, I don't think pre-allocating the buckets would be that costly as those are indeed empty COW array lists, so I guess I could do that instead of using atomic ref arrays.

Reply:
I think there will be performance savings with final arrays and preallocated buckets, with little cost.
I agree that we should punt log arrays to a different issue.

Reply:
Agreed, I've changed the implementation to use pre-allocated `Pool` arrays instead.